### PR TITLE
Add spanner support as an SQL output

### PR DESCRIPTION
**Pull request by rvilim**
This adds Google Cloud Spanner (https://cloud.google.com/spanner?hl=en) support for sql_raw and sql_insert. 

Since spanner conforms to the standards sql library this is mostly just an import. Quoting in spanner sql is slightly different than normal sql (it doesn't like double quotes) so some conditional logic had to be added, particularly in the integration tests.

The integration tests are done in the same fashion as all the other ones, with a docker container that runs the spanner emulator (dockerfile + startup script included here)

## Note
This is a resubmission of #122 which I am closing. #122 removes some uneeded db.Close() calls, we decided that's a bit scary. It was honestly eaiser just to make a new branch opting in to the changes I need rather than undoing all the changes I don't need.

Here we re-instantiate the DB object in the tests a bunch because spanner will not let you reuse one once you've closed it.

### CODE REVIEW

File: internal/impl/sql/integration_test.go [hunk 1]
```diff
@@ -718,6 +758,78 @@ func TestIntegrationPostgres(t *testing.T) {
...
+  baz string(50) not null,
+		) primary key (foo)`, name))
+		return name, err
+	}
+
+	require.NoError(t, pool.Retry(func() error {
+		var db *sql.DB
+		db, err = sql.Open("spanner", dsn)
+		if err != nil {
+			fmt.Println(`open error`, err)
```
Comment (jem-davies) [inline internal/impl/sql/integration_test.go:815]:
```suggestion
			t.Logf(`open error: %s`, err)
```

File: internal/impl/sql/integration_test.go [hunk 2]
```diff
@@ -718,6 +758,78 @@ func TestIntegrationPostgres(t *testing.T) {
...
+			return err
+		}
+		if err = db.Ping(); err != nil {
+			fmt.Println(`ping error`, err)
```
Comment (jem-davies) [inline internal/impl/sql/integration_test.go:819]:
```suggestion
			t.Logf(`ping error: %s`, err)
```

File: internal/impl/sql/integration_test.go [hunk 3]
```diff
@@ -718,6 +758,78 @@ func TestIntegrationPostgres(t *testing.T) {
...
+			db.Close()
+			return err
+		}
+		if _, err := createTable("footable"); err != nil {
+			fmt.Println(`create table error`, err)
```
Comment (jem-davies) [inline internal/impl/sql/integration_test.go:824]:
```suggestion
			t.Logf(`create table error: %s`, err)
```

File: internal/impl/sql/integration_test.go [hunk 4]
```diff
@@ -718,6 +758,78 @@ func TestIntegrationPostgres(t *testing.T) {
...
 }
 
+func TestIntegrationSpanner(t *testing.T) {
+	integration.CheckSkip(t)
+
+	pool, err := dockertest.NewPool("")
+	if err != nil {
+		t.Skipf("Could not connect to docker: %s", err)
+	}
+	pool.MaxWait = 10 * time.Second
```
Comment (jem-davies) [inline internal/impl/sql/integration_test.go:768]:
```suggestion
	pool.MaxWait = 3 * time.Minute
```

just so it's the same as the other tests in this file

File: internal/impl/sql/integration_test.go [hunk 5]
```diff
@@ -18,6 +18,8 @@ import (
 	"github.com/warpstreamlabs/bento/public/service"
 	"github.com/warpstreamlabs/bento/public/service/integration"
 
+	_ "github.com/googleapis/go-sql-spanner"
+
```
Comment (jem-davies) [inline internal/impl/sql/integration_test.go:22]:
```suggestion

```

Comment (jem-davies) [reply internal/impl/sql/integration_test.go:22]: not needed?

File: internal/impl/sql/integration_test.go [hunk 6]
```diff
@@ -718,6 +758,78 @@ func TestIntegrationPostgres(t *testing.T) {
...
+	// This needs to be set so that the Spanner connector will pick up that we are using the emulator.
+	t.Setenv("SPANNER_EMULATOR_HOST", emulatorHost)
+
+	project := "test-project"
+	instance := "test-instance"
+	database := "test-database"
+
+	dsn := fmt.Sprintf("projects/%s/instances/%s/databases/%s", project, instance, database)
+
+	t.Cleanup(func() {
```
Comment (jem-davies) [inline internal/impl/sql/integration_test.go:793]: should we add db.Close() to this cleanup function?

**Review by jem-davies — COMMENTED** [review]
just a few small things - looks good 👍  thanks a lot for the contribution

File: internal/impl/sql/integration_test.go [hunk 7]
```diff
@@ -718,6 +758,78 @@ func TestIntegrationPostgres(t *testing.T) {
...
+
+	resource, err := pool.BuildAndRunWithBuildOptions(&dockertest.BuildOptions{
+		ContextDir: "./resources",
+		Dockerfile: "Dockerfile_spanner",
+	}, &dockertest.RunOptions{
+		Name:         "spannertest",
+		ExposedPorts: []string{"9010/tcp", "9020/tcp"},
+	})
+	if err != nil {
+		fmt.Printf("Could not start resource: %s", err)
```
Comment (jem-davies) [inline internal/impl/sql/integration_test.go:778]:
```suggestion
		t.Logf("Could not start resource: %s", err)
```

File: internal/impl/sql/integration_test.go (hunk 5 above)
Comment (rvilim) [reply internal/impl/sql/integration_test.go:22]: oh oops! thanks

File: internal/impl/sql/integration_test.go (hunk 6 above)
Comment (rvilim) [reply internal/impl/sql/integration_test.go:791]: I'm not sure we can? There is no db object in scope there. Also, at that point the entire docker container is getting killed so to a certain extent it doesn't matter.

Comment (rvilim) [reply internal/impl/sql/integration_test.go:791]: There isn't a db object in scope in `Cleanup` atm. Or do you mean adding


```go
var db *sql.DB
db, err = sql.Open("spanner", dsn)
db.Close()
```

at the start of cleanup? Happy to do that if you like.

**Comment by rvilim** [comment]
> just a few small things - looks good 👍 thanks a lot for the contribution

Thanks for catching my silliness! I had one q about what you were looking for in the cleanup function. Happy to go either way, I'm not sure it matters a ton since we are killing the docker container and everything is going out of scope anyways.

File: internal/impl/sql/integration_test.go (hunk 6 above)
Comment (jem-davies) [reply internal/impl/sql/integration_test.go:791]: yeah would need to do that and alter the line below to := ... Can see how it's done in the other tests.

**Pull request merged**